--- ./asn.1/asn_internal2.h	2022-11-29 13:56:23.000000000 +0100
+++ ./asn.1/asn_internal.h	2022-11-29 13:57:28.000000000 +0100
@@ -102,7 +102,12 @@
 /*
  * Check stack against overflow, if limit is set.
  */
-#define	ASN__DEFAULT_STACK_MAX	(30000)
+#if defined(__APPLE__) && defined(DEBUG)
+/* Assume address sanitizer is enabled, which will blow up the stack a lot */
+#define ASN__DEFAULT_STACK_MAX (3000000)
+#else
+#define ASN__DEFAULT_STACK_MAX (30000)
+#endif
 static int __attribute__((unused))
 ASN__STACK_OVERFLOW_CHECK(asn_codec_ctx_t *ctx) {
 	if(ctx && ctx->max_stack_size) {
